### PR TITLE
Update table.grow and memory.grow reduction rule

### PR DESCRIPTION
**Pull request by ShinWonho**
As discussed in the [discussion](https://github.com/orgs/Wasm-DSL/discussions/29), I change instance definition to have type, and add typing in the reduction rule for checking upper limit.

### CODE REVIEW

File: spectec/spec/4-runtime.watsup [hunk 1]
```diff
@@ -52,13 +52,25 @@ def $default_(EXTERNREF) = (REF.NULL EXTERNREF) ;; TODO: All reference types sho
...
+  { TYPE globaltype,
+    VALUE val }
+syntax tableinst hint(desc "table instance") =
+  { TYPE tabletype,
+    ELEM ref* }
+syntax meminst hint(desc "memory instance") =
+  { TYPE memtype,
+    DATA byte* }
+syntax eleminst hint(desc "element instance") =
+  { ELEM ref* }
```
[the remark this answers is not in the document]
Comment (ShinWonho) [reply spectec/spec/4-runtime.watsup]: Sure, I added the `TYPE` field in `eleminst`

File: spectec/spec/6-reduction.watsup [hunk 2]
```diff
@@ -172,44 +173,46 @@ rule Step_pure/local.tee:
...
   z; (TABLE.SIZE x)  ~>  (CONST I32 n)
-  -- if |$table(z, x)| = n  ;; TODO: inline this
+  -- if |$table(z, x).ELEM| = n  ;; TODO: inline this
 
 
 rule Step/table.grow-succeed:
-  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x)|))
+  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x).ELEM|))
+  -- if $table(z, 0).TYPE = tabletype
+  -- Tabletype_ok: |- tabletype : OK
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:207]: Hm, AFAICS, that would just revalidate the existing table type, which should already be valid.

What is needed here are three steps: (1) modifying the table type to the new lower limit, (2) checking that these new limits are still valid, if okay, update the table's actual type to the new type.

File: spectec/spec/6-reduction.watsup [hunk 3]
```diff
@@ -302,19 +305,21 @@ rule Step/store-pack-val:
...
   z; (MEMORY.SIZE)  ~>  (CONST I32 n)
-  -- if $(n * 64 * $Ki) = |$mem(z, 0)|
+  -- if $(n * 64 * $Ki) = |$mem(z, 0).DATA|
 
 
 rule Step/memory.grow-succeed:
-  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_memext(z, 0, 0^(n * 64 * $Ki)); (CONST I32 $(|$mem(z, 0)| / (64 * $Ki)))
+  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_memext(z, 0, 0^(n * 64 * $Ki)); (CONST I32 $(|$mem(z, 0).DATA| / (64 * $Ki)))
+  -- if $mem(z, 0).TYPE = memtype
+  -- Memtype_ok: |- memtype : OK
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:314]: Same here.

File: spectec/spec/6-reduction.watsup (hunk 2 above)
Comment (ShinWonho) [reply spectec/spec/6-reduction.watsup]: Oh, thank you for the comment. I fixed it.

Comment (rossberg) [reply spectec/spec/6-reduction.watsup]: I'm afraid this hasn't fixed it, it's still just revalidating the old type unchanged.

Growing a table/memory updates the lower limit of its instance's dynamic type to the new size. And the validity check on the type is one way to ensure that this doesn't violate the upper limit.

**Comment by ShinWonho** [comment]
I fixed rules and checked that it works well in the interpreter.
(These pull request depends on #30 )
Now test pass rate is 100%!

File: spectec/spec/1-syntax.watsup [hunk 4]
```diff
@@ -57,7 +57,7 @@ syntax resulttype hint(desc "result type") =
   valtype*
 
 syntax limits hint(desc "limits") =
-  `[u32 .. u32]
+  `[u32 .. u32?]
```
Comment (rossberg) [inline spectec/spec/1-syntax.watsup:60]: See comment on other PR, I think we could leave this as is.

File: spectec/spec/6-reduction.watsup [hunk 5]
```diff
@@ -172,44 +173,47 @@ rule Step_pure/local.tee:
...
-  -- if |$table(z, x)| = n  ;; TODO: inline this
+  -- if |$table(z, x).ELEM| = n  ;; TODO: inline this
 
 
 rule Step/table.grow-succeed:
-  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x)|))
+  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableinst(z, x, ti'); (CONST I32 $(|$table(z, x).ELEM|))
+  -- if $table(z, x) = ti
+  -- if $grow_table(ti, n, ref) = ti'
+  -- Tabletype_ok: |- ti'.TYPE : OK
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:208]: Can we move this inside $grow_table?

File: spectec/spec/4-runtime.watsup [hunk 6]
```diff
@@ -123,23 +136,33 @@ def $local((s; f), x) = f.LOCAL[x]
...
+def $with_meminst((s; f), x, mi) = s[.MEM[f.MODULE.MEM[x]] = mi]; f
+def $with_elem((s; f), x, r*) = s[.ELEM[f.MODULE.ELEM[x]].ELEM = r*]; f
+def $with_data((s; f), x, b*) = s[.DATA[f.MODULE.DATA[x]].DATA = b*]; f
+
+def $grow_table(tableinst, nat, ref) : tableinst
+def $grow_memory(meminst, nat) : meminst
+def $grow_table(ti, n, r) = ti'
+  -- if ti = { TYPE `[i .. j?] reftype, ELEM r'* }
+  -- if i' = $(|r'*| + n)
+  -- if ti' = { TYPE `[i' .. j?] reftype, ELEM r'* r^n }
```
Comment (rossberg) [inline spectec/spec/4-runtime.watsup:161]:
```suggestion
  -- if ti = { TYPE `[i .. j?] rt, ELEM r'* }
  -- if i' = $(|r'*| + n)
  -- if ti' = { TYPE `[i' .. j?] rt, ELEM r'* r^n }
```

File: spectec/spec/6-reduction.watsup [hunk 7]
```diff
@@ -302,19 +306,22 @@ rule Step/store-pack-val:
...
-  -- if $(n * 64 * $Ki) = |$mem(z, 0)|
+  -- if $(n * 64 * $Ki) = |$mem(z, 0).DATA|
 
 
 rule Step/memory.grow-succeed:
-  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_memext(z, 0, 0^(n * 64 * $Ki)); (CONST I32 $(|$mem(z, 0)| / (64 * $Ki)))
+  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_meminst(z, 0, mi'); (CONST I32 $(|$mem(z, 0).DATA| / (64 * $Ki)))
+  -- if $mem(z, 0) = mi
+  -- if $grow_memory(mi, n) = mi'
+  -- Memtype_ok: |- mi'.TYPE : OK
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:316]: Same here.

File: spectec/spec/6-reduction.watsup [hunk 8]
```diff
@@ -302,19 +306,22 @@ rule Step/store-pack-val:
...
   z; (MEMORY.SIZE)  ~>  (CONST I32 n)
-  -- if $(n * 64 * $Ki) = |$mem(z, 0)|
+  -- if $(n * 64 * $Ki) = |$mem(z, 0).DATA|
 
 
 rule Step/memory.grow-succeed:
-  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_memext(z, 0, 0^(n * 64 * $Ki)); (CONST I32 $(|$mem(z, 0)| / (64 * $Ki)))
+  z; (CONST I32 n) (MEMORY.GROW)  ~>  $with_meminst(z, 0, mi'); (CONST I32 $(|$mem(z, 0).DATA| / (64 * $Ki)))
+  -- if $mem(z, 0) = mi
+  -- if $grow_memory(mi, n) = mi'
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:315]:
```suggestion
  -- if $grow_memory($mem(z, 0), n) = mi'
```

File: spectec/spec/6-reduction.watsup [hunk 9]
```diff
@@ -172,44 +173,47 @@ rule Step_pure/local.tee:
...
   z; (TABLE.SIZE x)  ~>  (CONST I32 n)
-  -- if |$table(z, x)| = n  ;; TODO: inline this
+  -- if |$table(z, x).ELEM| = n  ;; TODO: inline this
 
 
 rule Step/table.grow-succeed:
-  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x)|))
+  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableinst(z, x, ti'); (CONST I32 $(|$table(z, x).ELEM|))
+  -- if $table(z, x) = ti
+  -- if $grow_table(ti, n, ref) = ti'
```
Comment (rossberg) [inline spectec/spec/6-reduction.watsup:207]:
```suggestion
  -- if $grow_table($table(z, x), n, ref) = ti'
```

**Review by rossberg — COMMENTED** [review]
Oops, sorry for the delay

**Comment by ShinWonho** [comment]
Thanks for the comments!
I updated it.

**Review by rossberg — APPROVED** [review]

**Pull request merged**